feat : Customer 회원 가입 기능 구현

yjjjwww opened the pull request:
### 변경사항
**AS-IS**
- Customer 엔티티 설정했습니다. Customer 데이터베이스에는 이메일, 비밀번호, 연락처, 적립금, 삭제여부를 저장합니다.
- 회원가입시 이미 등록된 이메일이 있는지 확인합니다.
- 회원가입시 비밀번호 유효성 검사, 연락처 형식 유효성 검사를 합니다.
- Customer 비밀번호는 암호화해서 데이터베이스에 저장합니다.
- 기본 적립금은 0원으로 저장됩니다.
- postman으로 API 테스트 완료했고, 테스트 코드로 성공할 경우와 Custom Error가 발생할 경우 올바른 응답이 나오는지 확인했습니다.

**TO-BE**
- Seller 회원 가입
- Customer, Seller 로그인

### 테스트
- [x] 테스트 코드
- [x] API 테스트

yjjjwww committed:
feat : [feature/customer-signup] Customer 회원 가입 기능 구현

yjjjwww committed:
test : [feature/customer-signup] Customer 회원 가입 기능 테스트 코드 작성

- 성공했을 때 테스트 코드와 실패 했을 경우 테스트 코드 작성(ALREADY_SIGNUP_EMAIL, INVALID_PASSWORD, INVALID_PHONE)
- Controller, Service 테스트 코드 작성

mingming-mentor commented on src/main/java/com/yjjjwww/yunmarket/customer/entity/Customer.java:
```diff
@@ -0,0 +1,35 @@
+
+  @Id
+  @GeneratedValue(strategy = GenerationType.IDENTITY)
+  Long id;
```
private을 넣어주심 좋을듯합니다

mingming-mentor commented on src/main/java/com/yjjjwww/yunmarket/customer/service/CustomerServiceImpl.java:
```diff
@@ -0,0 +1,85 @@
+
+  private final CustomerRepository customerRepository;
+
+  private static final String SIGNUP_SUCCESS = "회원가입 성공";
```
이 부분은 controller (view)에서 필요한 정보라고 보여요. 서비스로직에서는 성공시 void 아무값을 return 안하고 controller에서 exception없이 service로직이 수행되면 위 string을 반환하는 형태로 구현하시면 좋을듯합니다

mingming-mentor commented on src/main/java/com/yjjjwww/yunmarket/customer/service/CustomerServiceImpl.java:
```diff
@@ -0,0 +1,85 @@
+    }
+    boolean err = false;
+    String regex = "^[0-9]*$";
+    Pattern p = Pattern.compile(regex);
```
Pattern은 만드는데 조금 비용이 드는코드라서 static 으로 상단에 선언해주시면 좋을듯해요

mingming-mentor commented on src/main/java/com/yjjjwww/yunmarket/customer/service/CustomerServiceImpl.java:
```diff
@@ -0,0 +1,85 @@
+    if (m.matches()) {
+      err = true;
+    }
+    return err;
```
`return m.matches()`로 코드를 좀더 가독성 높게 작성해볼수 있을 것 같아요

mingming-mentor commented on src/main/java/com/yjjjwww/yunmarket/customer/service/CustomerServiceImpl.java:
```diff
@@ -0,0 +1,85 @@
+   * 핸드폰 번호 형식 검사
+   */
+  private static boolean isValidPhone(String phone) {
+    if (phone.length() > 11 || phone.length() < 5) {
```
요렇게 하셔도 되지만 request에서 @Valid를 걸어서 어노테이션으로 처리해도 될듯해요

mingming-mentor commented on src/main/java/com/yjjjwww/yunmarket/customer/service/CustomerServiceImpl.java:
```diff
@@ -0,0 +1,85 @@
+   */
+  private static boolean isValidPassword(String password) {
+    boolean err = false;
+    String regex = "^(?=.*[A-Za-z])(?=.*\\d)(?=.*[$@$!%*#?&])[A-Za-z\\d$@$!%*#?&]{8,}$";
```
이 부분도 위 코멘트 참고해주세요

mingming-mentor commented on src/main/resources/application.properties:
```diff
@@ -1 +1,7 @@
+spring.datasource.driver-class-name=com.mysql.cj.jdbc.Driver
+spring.datasource.url=jdbc:mysql://localhost:3306/yunmarket?characterEncoding=UTF-8&serverTimezone=UTC
+spring.datasource.username=root
+spring.datasource.password=yjw202020@
```
요런 정보는 public으로 올리는 것보다 github의 secret 기능을 이용해보셔도 좋을 것 같아요

mingming-mentor commented on src/main/java/com/yjjjwww/yunmarket/customer/controller/CustomerController.java:
```diff
@@ -0,0 +1,23 @@
+  private final CustomerService customerService;
+
+  @PostMapping("/signUp")
+  public ResponseEntity<String> signUp(@RequestBody CustomerSignUpForm customerSignUpForm) {
```
다른 스터디원분 코드에서 드렸던 리뷰인데 동일하게 드려볼게요
* * *
service에서 사용하는 dto와 controller에서 받아오는 request는 분리하는 것이 좋습니다. 지금은 controller에서 하나의 service만 호출하긴 하지만 여러개의 service 로직을 호출할때도 있기 때문에 controller단과 service를 분리해주시면 좋아요. 참고하실만한 블로그 링크도 같이 드려볼게요  
https://techblog.woowahan.com/2711/ 의 2. Controller와 Service 레이어의 강한 결합 참고해주세요

mingming-mentor commented on src/main/java/com/yjjjwww/yunmarket/customer/service/CustomerServiceImpl.java:
```diff
@@ -0,0 +1,85 @@
+    if (optionalCustomer.isPresent()) {
+      throw (new CustomException(ErrorCode.ALREADY_SIGNUP_EMAIL));
+    }
+
```
요 부분도 요렇게 작성해볼수 있을 것 같아요
~~~java
customerRepository.findByEmail(customerSignUpForm.getEmail()).ifPresent(it -> {
    throw new CustomException(ErrorCode.ALREADY_SIGNUP_EMAIL)
});
~~~

mingming-mentor reviewed (approved):
수고 많으셨어요. 몇가지 리뷰 드려보았습니다. 
code convention 적용하셨는지 한번 확인부탁드릴게요. tab이 space 2개로 되어있는듯 하네요. 작성하신뒤 intelliJ에서 mac 기준으로 control + command + L 입력부탁드립니다 (reformat the code 기능 이용하시면되어요)  
수고 많으셨습니다

yjjjwww committed:
refactor : [feature/customer-signup] Customer 회원 가입 기능 코드 개선

- Customer 엔티티 id에 private 추가
- CustomerService signUp 메서드를 void 형태로 변경하고 Controller에서 exception없이 Service 로직이 수행되면 응답으로 "회원가입 성공"을 반환
- Password, Phone 유효성 검사에서 Pattern 문자열을 static으로 상단에 선언, return 값 형태 간결화
- Controller에서 받는 dto와 Service에서 사용하는 dto를 분리
- 같은 이메일의 Customer가 있는지 확인하는 부분 코드 리팩토링
